File: packages/plugin/internal/rename/title_engine_test.go
package rename

import (
	"slices"
	"strings"
	"testing"
)

func TestRenamesGenericTitleAfterMeaningfulPrompt(t *testing.T) {
	config := DefaultConfig()
	config.Mode = ModeDryRun
	state := InitialThreadState("thread-1")
	cwd := "/tmp/project"
	prompt := CreatePendingPrompt(
		"turn-1",
		"Build a Bun TypeScript Codex hook that auto renames thread titles",
		&cwd,
	)
	title := "New chat"

	decision := DecideRename(DecideInput{
		Config:           config,
		State:            &state,
		Thread:           &ThreadRecord{ID: "thread-1", Name: &title},
		Prompt:           &prompt,
		AssistantMessage: "Implemented the package and tests.",
		Transcript:       TranscriptTail{ToolSignals: []string{"edited files"}},
		SourceHash:       "source-1",
	})

	if decision.Kind != "would_rename" {
		t.Fatalf("expected would_rename, got %s", decision.Kind)
	}
	if decision.NewTitle == nil {
		t.Fatal("expected new title")
	}
	if !strings.Contains(*decision.NewTitle, "Codex") || !strings.Contains(*decision.NewTitle, "Auto") {
		t.Fatalf("expected Codex Auto title, got %q", *decision.NewTitle)
	}
}

func TestSkipsSameTaskFollowUpUnderCooldown(t *testing.T) {
	config := DefaultConfig()
	config.Mode = ModeDryRun
	state := InitialThreadState("thread-1")
	cwd := "/tmp/project"
	firstPrompt := CreatePendingPrompt("turn-1", "Research Codex auto rename hooks", &cwd)
	title := "New chat"

	firstDecision := DecideRename(DecideInput{
		Config:           config,
		State:            &state,
		Thread:           &ThreadRecord{ID: "thread-1", Name: &title},
		Prompt:           &firstPrompt,
		AssistantMessage: "Researched Codex hooks and app server thread naming.",
		Transcript:       TranscriptTail{},
		SourceHash:       "source-1",
	})
	nextState := ApplyDecisionToState(state, firstDecision)
	if firstDecision.NewTitle == nil {
		t.Fatal("expected first decision to produce a title")
	}

	followUp := CreatePendingPrompt("turn-2", "Can you add tests for that same hook?", &cwd)
	followUpTitle := *firstDecision.NewTitle
	followUpDecision := DecideRename(DecideInput{
		Config:           config,
		State:            &nextState,
		Thread:           &ThreadRecord{ID: "thread-1", Name: &followUpTitle},
		Prompt:           &followUp,
		AssistantMessage: "Added tests for the same hook.",
		Transcript:       TranscriptTail{ToolSignals: []string{"tests"}},
		SourceHash:       "source-2",
	})

	if followUpDecision.Kind != "skipped" {
		t.Fatalf("expected skipped, got %s", followUpDecision.Kind)
	}
}

func TestDetectsManualTitleOverride(t *testing.T) {
	config := DefaultConfig()
	config.Mode = ModeDryRun
	config.RespectManualTitles = true
	state := InitialThreadState("thread-1")
	lastAutoTitle := "Codex Auto Rename Hook"
	state.LastAutoTitle = &lastAutoTitle
	cwd := "/tmp/project"
	intent := CreatePendingPrompt("turn-1", "Codex auto rename hook", &cwd).Intent
	state.StableIntent = &intent
	prompt := CreatePendingPrompt("turn-2", "Switch gears and research mobile SQLCipher boot errors", &cwd)
	manualTitle := "My Hand Picked Title"

	decision := DecideRename(DecideInput{
		Config:           config,
		State:            &state,
		Thread:           &ThreadRecord{ID: "thread-1", Name: &manualTitle},
		Prompt:           &prompt,
		AssistantMessage: "Researched SQLCipher boot handling.",
		Transcript:       TranscriptTail{},
		SourceHash:       "source-2",
	})

	if decision.Kind != "skipped" {
		t.Fatalf("expected skipped, got %s", decision.Kind)
	}
	if !slices.Contains(decision.Signals, "manual_title_detected") {
		t.Fatalf("expected manual_title_detected signal, got %#v", decision.Signals)
	}
}

File: .old/tests/title-engine.test.ts
import { describe, expect, test } from 'bun:test'

import { defaultConfig } from '../src/config.ts'
import { initialThreadState } from '../src/state.ts'
import { applyDecisionToState, createPendingPrompt, decideRename } from '../src/title-engine.ts'

describe('title engine', () => {
	test('renames a generic title after a meaningful prompt', async () => {
		const config = { ...defaultConfig(), mode: 'dry-run' as const }
		const state = initialThreadState('thread-1')
		const prompt = createPendingPrompt(
			'turn-1',
			'Build a Bun TypeScript Codex hook that auto renames thread titles',
			'/tmp/project'
		)
		const decision = await decideRename({
			config,
			state,
			thread: { id: 'thread-1', name: 'New chat' },
			prompt,
			assistantMessage: 'Implemented the package and tests.',
			transcript: { userMessages: [], assistantMessages: [], toolSignals: ['edited files'] },
			sourceHash: 'source-1'
		})
		expect(decision.kind).toBe('would_rename')
		expect(decision.newTitle).toContain('Codex')
		expect(decision.newTitle).toContain('Auto')
	})

	test('skips same-task follow-up under cooldown', async () => {
		const config = { ...defaultConfig(), mode: 'dry-run' as const }
		const state = initialThreadState('thread-1')
		const firstPrompt = createPendingPrompt(
			'turn-1',
			'Research Codex auto rename hooks',
			'/tmp/project'
		)
		const firstDecision = await decideRename({
			config,
			state,
			thread: { id: 'thread-1', name: 'New chat' },
			prompt: firstPrompt,
			assistantMessage: 'Researched Codex hooks and app server thread naming.',
			transcript: { userMessages: [], assistantMessages: [], toolSignals: [] },
			sourceHash: 'source-1'
		})
		const nextState = applyDecisionToState(state, firstDecision)
		const followUp = createPendingPrompt(
			'turn-2',
			'Can you add tests for that same hook?',
			'/tmp/project'
		)
		const followUpDecision = await decideRename({
			config,
			state: nextState,
			thread: { id: 'thread-1', name: firstDecision.newTitle },
			prompt: followUp,
			assistantMessage: 'Added tests for the same hook.',
			transcript: { userMessages: [], assistantMessages: [], toolSignals: ['tests'] },
			sourceHash: 'source-2'
		})
		expect(followUpDecision.kind).toBe('skipped')
	})

	test('detects manual title override', async () => {
		const config = { ...defaultConfig(), mode: 'dry-run' as const, respectManualTitles: true }
		const state = {
			...initialThreadState('thread-1'),
			lastAutoTitle: 'Codex Auto Rename Hook',
			stableIntent: createPendingPrompt('turn-1', 'Codex auto rename hook', '/tmp/project').intent
		}
		const prompt = createPendingPrompt(
			'turn-2',
			'Switch gears and research mobile SQLCipher boot errors',
			'/tmp/project'
		)
		const decision = await decideRename({
			config,
			state,
			thread: { id: 'thread-1', name: 'My Hand Picked Title' },
			prompt,
			assistantMessage: 'Researched SQLCipher boot handling.',
			transcript: { userMessages: [], assistantMessages: [], toolSignals: [] },
			sourceHash: 'source-2'
		})
		expect(decision.kind).toBe('skipped')
		expect(decision.signals).toContain('manual_title_detected')
	})
})
